src/subclass/input_stream.rs

```diff
@@ -0,0 +1,396 @@
+                Err(from_glib_full(err))
+            } else {
+                Ok(res as usize)
+            }
+        }
+    }
+
+    fn parent_close(
+        &self,
+        stream: &InputStream,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<(), Error> {
+        unsafe {
+            let data = self.get_type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gio_sys::GInputStreamClass;
+            let mut err = ptr::null_mut();
+            if let Some(f) = (*parent_class).close_fn {
+                if from_glib(f(
+                    stream.to_glib_none().0,
+                    cancellable.to_glib_none().0,
+                    &mut err,
+                )) {
+                    Ok(())
+                } else {
+                    Err(from_glib_full(err))
+                }
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    fn parent_skip(
+        &self,
+        stream: &InputStream,
+        count: usize,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<usize, Error> {
+        unsafe {
+            let data = self.get_type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gio_sys::GInputStreamClass;
+            let mut err = ptr::null_mut();
+            let f = (*parent_class)
+                .skip
+                .expect("No parent class implementation for \"skip\"");
+            let res = f(
+                stream.to_glib_none().0,
+                count,
+                cancellable.to_glib_none().0,
+                &mut err,
+            );
+            if res == -1 {
+                Err(from_glib_full(err))
+            } else {
+                Ok(res as usize)
+            }
+        }
+    }
+}
+
+unsafe impl<T: ObjectSubclass + InputStreamImpl> IsSubclassable<T> for InputStreamClass {
+    fn override_vfuncs(&mut self) {
+        <glib::ObjectClass as IsSubclassable<T>>::override_vfuncs(self);
+        unsafe {
+            let klass = &mut *(self as *mut Self as *mut gio_sys::GInputStreamClass);
+            klass.read_fn = Some(stream_read::<T>);
+            klass.close_fn = Some(stream_close::<T>);
+            klass.skip = Some(stream_skip::<T>);
+        }
+    }
+}
+
+unsafe extern "C" fn stream_read<T: ObjectSubclass>(
+    ptr: *mut gio_sys::GInputStream,
+    buffer: glib_sys::gpointer,
+    count: usize,
+    cancellable: *mut gio_sys::GCancellable,
+    err: *mut *mut glib_sys::GError,
+) -> isize
+where
+    T: InputStreamImpl,
+{
+    use std::slice;
+
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: InputStream = from_glib_borrow(ptr);
+
+    match imp.read(
+        &wrap,
+        slice::from_raw_parts_mut(buffer as *mut u8, count),
+        Option::<Cancellable>::from_glib_borrow(cancellable).as_ref(),
+    ) {
+        Ok(res) => {
+            use std::isize;
+            assert!(res < isize::MAX as usize);
```

Comment:
Not a big fan of `assert` calls. :-/

Reply:
If we don't assert here, then C will assert one step further :) It's broken API at the C level.

And more important, this case can't ever happen at all: before calling `stream_read()`, C will check that the `count` is smaller than `isize::MAX`. We will never ever get called with a bigger value than that, so if the subclass reads more than there is memory available then this is a serious bug in the subclass that we can't do anything other than panicking.

You prefer segfaults over panics? :)

Reply:
But thanks for reminding me to add some more assertions here to ensure things don't fall apart.

Reply:
No, I prefer `Result`s over panics. :3

Reply:
We have this discussion over and over again. There's a difference between recoverable/handleable errors (-> `Result`) and fatal errors (-> panics). This is the second category :)

Reply:
I know but I need to make my point. \*anti-panic instinct\*

Reply:
Added some more assertions to make things more explicit. All these parts are also checked on the C side and causing assertions there, having them on our side only makes these constraints more clear.